### PR TITLE
Feat/vip outbound

### LINKED ISSUE #765: Pod converter for Kuma IP Outbound
### Summary

In #764, we suggest to implement a dedicated Kuma CP DNS resolver, allocating virtual IP addresses for the Kuma controlled services in the `.kuma` zone. In this feature request, we specify how these virtual IPs are leveraged.

Two options to implement:
1) The Outbound Pod converter in [outbound_converter.go](https://github.com/Kong/kuma/blob/master/pkg/plugins/discovery/k8s/controllers/outbound_converter.go), will check the services in the DNS resolver and created Outbound listeners for the corresponding IPs.

2) Update the data plane resources with the Outbound generated for that service

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

Implement the VIP Outbound generation.

### Issues resolved

Fix #765 

### Documentation

- [ ] TBD Link to the website [documentation PR](https://github.com/Kong/kuma-website/pull/XXX)

### CODE REVIEW

File: pkg/xds/server/components.go [hunk 1]
```diff
@@ -157,6 +154,14 @@ func DefaultDataplaneSyncTracker(rt core_runtime.Runtime, reconciler, ingressRec
 					return ingressReconciler.Reconcile(envoyCtx, &proxy)
 				}
 
+				// Generate VIP outbounds only when in Kubernetes
+				if rt.Config().Environment == config_core.KubernetesEnvironment {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/components.go:158]: check `dataplane.Spec.GetTransparentProxy() != nil` instead

File: pkg/xds/topology/vip_outbounds.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	dataplanes *mesh_core.DataplaneResourceList, resolver resolver.DNSResolver) (errs error) {
+	serviceVIPMap := map[string]string{}
+	for _, dp := range dataplanes.Items {
+		if dp.Meta.GetName() == dataplane.Meta.GetName() {
+			continue
+		}
+
+		for _, inbound := range dp.Spec.Networking.Inbound {
+			inService := inbound.GetTags()[mesh_proto.ServiceTag]
+
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/vip_outbounds.go:26]: check if `inService` is in map already and skip if so.

File: pkg/xds/topology/vip_outbounds.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+			vip, err := resolver.ForwardLookup(inService)
+			if err != nil {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/vip_outbounds.go:28]: This logic should be gone after we change `service.namespace.svc:port` to `service_namespace_svc_port` right?
Can you add todo to remove it?

File: pkg/xds/topology/vip_outbounds.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+				vip, err = resolver.ForwardLookup(split[0])
+				if err != nil {
+					errs = multierr.Append(errs, errors.Wrapf(err, "unable to resolve %s", inService))
+				}
+			}
+			serviceVIPMap[inService] = vip
+		}
+	}
+
+	for service, vip := range serviceVIPMap {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/vip_outbounds.go:40]: We need to sort the map before generating outbounds.

XDS is sent if the snapshot changes (compared by deepequal on protos), in this case different order means different snapshot.

File: pkg/xds/topology/vip_outbounds_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		}
+
+		// when
+		err = topology.PatchDataplaneWithVIPOutbounds(dataplane, &dataplanes, resolver)
+		// then
+		Expect(err).ToNot(HaveOccurred())
+		// and
+		Expect(len(dataplane.Spec.Networking.Outbound)).To(Equal(4))
+		// and
+		Expect(dataplane.Spec.Networking.Outbound[3].Port).To(Equal(topology.VIPListenPort))
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/vip_outbounds_test.go:85]: once you have order, you can also add assertions for services

File: pkg/xds/server/components.go (hunk 1 above)
Comment (nickolaev) [reply pkg/xds/server/components.go:158]: Also when no Ingress :)

File: pkg/xds/topology/vip_outbounds.go (hunk 2 above)
Comment (nickolaev) [reply pkg/xds/topology/vip_outbounds.go:28]: Done, thanks!

File: pkg/xds/topology/vip_outbounds.go (hunk 4 above)
Comment (nickolaev) [reply pkg/xds/topology/vip_outbounds.go]: Done, thanks!

File: pkg/xds/topology/vip_outbounds_test.go (hunk 5 above)
Comment (nickolaev) [reply pkg/xds/topology/vip_outbounds_test.go:89]: Done, thanks!

**Pull request merged**